src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/PreprocessingSymbolReferenceFinder.cs

```diff
@@ -0,0 +1,141 @@
+            cancellationToken).ConfigureAwait(false);
+
+        return normalReferences;
+
+        static bool MatchesPreprocessingReference(SyntaxToken token, FindReferencesDocumentState state)
+        {
+            var syntaxFacts = state.SyntaxFacts;
+
+            var tokenParent = token.Parent;
+            Debug.Assert(tokenParent is not null);
+
+            // Quickly evaluate the common case that the parent is a #define or #undef directive
+            var parentKind = tokenParent.RawKind;
+            if (parentKind == syntaxFacts.SyntaxKinds.DefineDirectiveTrivia)
+                return true;
+
+            if (parentKind == syntaxFacts.SyntaxKinds.UndefDirectiveTrivia)
+                return true;
+
+            // In VB, a #Const directive assigns a value that can also
+            // derive from preprocessing symbols
+            if (state.Document.Project.Language is LanguageNames.VisualBasic)
+            {
+                var extendedTokenParent = tokenParent;
+                while (true)
+                {
+                    extendedTokenParent = extendedTokenParent.Parent;
+
+                    if (extendedTokenParent is null)
+                        break;
+
+                    if (extendedTokenParent.RawKind == syntaxFacts.SyntaxKinds.DefineDirectiveTrivia)
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            // Otherwise, only inside an #if or #elif directive are preprocessing symbols used
+            return syntaxFacts.SpansIfOrElseIfPreprocessorDirective(tokenParent);
+        }
+    }
+
+    protected override bool CanFind(IPreprocessingSymbol symbol) => true;
+
+    protected override async Task<ImmutableArray<Document>> DetermineDocumentsToSearchAsync(
+        IPreprocessingSymbol symbol,
+        HashSet<string>? globalAliases,
+        Project project,
+        IImmutableSet<Document>? documents,
+        FindReferencesSearchOptions options,
+        CancellationToken cancellationToken)
+    {
+        // NOTE: We intentionally search for all documents in the entire solution. This is because
+        //       the symbols are validly bound by their requested name, despite their current definition
+        //       state. Therefore, the same symbol name could be shared across multiple projects and
+        //       configured in the project configuration with the same shared identifier.
+
+        return await FindAllSolutionDocumentsAsync(project.Solution, HasDirectiveProbablyContainsIdentifier, symbol, cancellationToken)
+            .ConfigureAwait(false);
+
+        static async ValueTask<bool> HasDirectiveProbablyContainsIdentifier(Document document, IPreprocessingSymbol symbol, CancellationToken ct)
+        {
+            var root = await document.GetSyntaxRootAsync(ct).ConfigureAwait(false);
+            if (root is not { ContainsDirectives: true })
```

Comment:
you're parsing all files.  You should just use the Syntax indices we already have to determine this.

Reply:
I do not know how this could be done; in this method there is no direct access to a syntax tree index, and the information is passed directly through the syntax root. No better alternative comes to my mind.

Reply:
there's a reference to the syntax tree index two lines below this :)

Reply:
I remember in our discussion that I would have to filter through the syntax root's index first for better performance, which contains this `ContainsDirective` property, before relying on the syntax tree index, where I had initially placed a new property specifically for pp directives. Which path sounds more preferable?

Reply:
Neither.  The index should just store if it contains directives in it.  That can be done trivially when teh index is built, as it literally is a bit on the root that can be checked immediately.

Reply:
Correct me if I'm wrong, but the syntax root's index and the syntax tree index are different. The syntax root's index is a classic node information index, using `NodeFlags`, where one of the bits is called `ContainsDirectives` which I am currently used on the parsed syntax tree. The syntax tree index is the one containing the `ProbablyContainsIdentifier` method that filters out documents without the given identifier. And this is why I resorted to this logic.

Reply:
> The syntax tree index is the one containing the ProbablyContainsIdentifier method that filters out documents without the given identifier. 

Yes.  Add the ContainsDirective piece of information to this latter index.  That way we can see if we need to use this file without parsing it. 

Right now, your code parses *every file* to look for pp directives.  Even if the file doesn't even contain directives.  That's a massive performance cost.  We don't want that.  We want to be able to filter, using the index, *only* to files that both contain this identifier and have pp directives in them.

Reply:
> Yes. Add the ContainsDirective piece of information to this latter index. That way we can see if we need to use this file without parsing it.

Ah so that's what you meant. Yes, will add it to the syntax tree index.